### PR TITLE
[BYOC] FTVMAnnotateTarget method signature update

### PR DESCRIPTION
**Pull request by d-smirnov**
This PR changes signature of `FTVMAnnotateTarget` to `runtime::TypedPackedFunc<bool(const Expr& expr)>`
which allows to utilise extra information (e.g. an operation output type) from passed expr argument.

### CODE REVIEW

**Review by comaniac — APPROVED** [review]
LGTM.
cc @zhiics @masahi 
Also cc @trevor-m @mbaret. Please help double check.

File: tests/python/contrib/test_arm_compute_lib/test_conv2d.py [hunk 1]
```diff
@@ -21,15 +21,15 @@
 import tvm
 from tvm import relay
 
-from .infrastructure import (
+from test_arm_compute_lib.infrastructure import (
```
Comment (mbaret) [inline tests/python/contrib/test_arm_compute_lib/test_conv2d.py:24]: It's not obvious to me why this change is related.

File: python/tvm/relay/op/contrib/arm_compute_lib.py [hunk 2]
```diff
@@ -201,38 +201,32 @@ def check_avg_pool2d(extract):
...
         pool = extract.args[0]
-        return avg_pool2d(pool.attrs, pool.args)
+        return avg_pool2d(pool)
 
     return [
         ("arm_compute_lib.conv2d", conv_pattern(), check_conv),
         ("arm_compute_lib.qnn_conv2d", qnn_conv_pattern(), check_qnn_conv),
         ("arm_compute_lib.dense", dense_pattern(), check_dense),
         ("arm_compute_lib.qnn_dense", qnn_dense_pattern(), check_qnn_dense),
-        ("arm_compute_lib.qnn_conv2d", qnn_conv_pattern(), check_qnn_conv),
```
Comment (mbaret) [inline python/tvm/relay/op/contrib/arm_compute_lib.py:216]: Unrelated?

File: python/tvm/relay/op/contrib/arm_compute_lib.py [hunk 3]
```diff
@@ -248,8 +242,11 @@ def conv2d(attrs, args):
     return True
 
 
-def qnn_conv2d(attrs, args):
+# TODO check why no annotation is here
```
Comment (mbaret) [inline python/tvm/relay/op/contrib/arm_compute_lib.py:245]: Could you elaborate on this TODO?

**Review by mbaret — CHANGES_REQUESTED** [review]
This change may allow us to simplify some logic for Ethos-N as well. I'll leave that for a follow-up patch though.

File: tests/python/contrib/test_arm_compute_lib/test_conv2d.py (hunk 1 above)
Comment (d-smirnov) [reply tests/python/contrib/test_arm_compute_lib/test_conv2d.py:24]: to be able run tests as python -m test_arm_compute_lib.test_*

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 2 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py:216]: There was a duplication here (L213)

File: tests/python/contrib/test_arm_compute_lib/test_conv2d.py (hunk 1 above)
Comment (mbaret) [reply tests/python/contrib/test_arm_compute_lib/test_conv2d.py:24]: I think this should be in a separate patch then - doesn't relate to the change to annotate target.

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 2 above)
Comment (mbaret) [reply python/tvm/relay/op/contrib/arm_compute_lib.py:216]: Again, doesn't relate to this change so should be in a separate fix-up patch.

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 3 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: Removed TODO comments

Comment (lhutton1) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: Just an fyi, the reason qnn.conv2d wasn't registered is because as an individual operator it isn't supported. It needs to be accompanied with a requantize op.

File: python/tvm/relay/op/contrib/arm_compute_lib.py [hunk 4]
```diff
@@ -248,8 +242,11 @@ def conv2d(attrs, args):
...
+@tvm.ir.register_op_attr("qnn.conv2d", "target.arm_compute_lib")
```
Comment (mbaret) [inline python/tvm/relay/op/contrib/arm_compute_lib.py:245]: This shouldn't be registered. It's only supported as part of a composite pattern.

File: python/tvm/relay/op/contrib/arm_compute_lib.py [hunk 5]
```diff
@@ -279,8 +277,11 @@ def dense(attrs, args):
     return True
 
 
-def qnn_dense(attrs, args):
+# TODO check why no annotation is here
+@tvm.ir.register_op_attr("qnn.dense", "target.arm_compute_lib")
```
Comment (mbaret) [inline python/tvm/relay/op/contrib/arm_compute_lib.py:279]: This shouldn't be registered either.

**Review by mbaret — CHANGES_REQUESTED** [review]
Some stuff has become registered when it shouldn't be.

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 3 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: What was the reason to have a pattern and an operator annotations for nn.dense and nn.conv2d then? Both patterns happily accepts single ops.

Comment (lhutton1) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: nn.dense and nn.conv2d will work as a single op, we have all the data we need to create the accompanying ACL layer. qnn.dense and qnn.conv2d are subtly different in that we also require a qnn.requantize op, this op allows us to extract the output quantization info in order to create the layer in ACL. Hope that makes sense?

Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: Still not clear. Currently patterns allow `nn.dense` and `nn.conv2d` as single ops without any additional operator annotation. However, there is a difference is in arguments: annotations accept all variants of arguments but patterns restrict second argument to constant. So if operator is a single operator we are more lenient to its arguments. Could you elaborate this discordance please?

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 2 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py:216]: updated

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 4 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: updated

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 5 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: updated

**Review by comaniac — APPROVED** [review]
LGTM.

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 3 above)
Comment (lhutton1) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: That's a good point, the single op annotation will not check whether certain inputs are constant. We should probably fix this in another pr.

Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: @lhutton1 Agree. All constant args should be changed to var in `nn.conv2d`, `qnn.conv2d`, `nn.dense`, `qnn.dense` and operator annotations removed for `nn.conv2d` and `nn.dense`. Correct?

Comment (lhutton1) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: The patterns use `is_constant` for a reason - if I recall correctly it allows us to access the data at compile time, I believe they are treated differently to vars. My thought process was to keep the patterns and annotations as they are and simply add checks in `nn.conv2d` and `nn.dense` to make sure that certain args i.e. kernel are indeed constant, if not we return false and the operator isn't annotated?

Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: @lhutton1 Could you point me where exactly on compile time we depend on constant type of the second argument of `nn.conv2d`, `qnn.conv2d`, `nn.dense`, or `qnn.dense`, please?

Comment (lhutton1) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: Its not too obvious because the JSON codegen module and metadata module take care of constant serialization, I can't remember exactly how this works. Here you can see a constant being added to the JSON representation: https://github.com/apache/incubator-tvm/blob/main/src/relay/backend/contrib/codegen_json/codegen_json.h#L250 . One example in ACL codegen is we convert the layout of constants to the correct format for ACL: https://github.com/apache/incubator-tvm/blob/main/src/relay/backend/contrib/arm_compute_lib/codegen.cc#L340, hope that helps?

Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: @lhutton1 Still not clear. Is the second arg needs to be of constant type just to eliminate the necessity to convert it on run time when JSON serialization (with actual data available) happens? But in case of nn.conv2d typical kernel size is small and sort of negligible. Am I correct here? Or some other considerations were used?

Comment (lhutton1) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: Apologies if I'm misunderstanding, so you want to allow the possibility of inputting weights at runtime? If that's the case, why wouldn't they be available at compile time? I think the only consideration was that since weights are normally constant we can optimize conversion to the correct layout ahead of time, avoiding the need to do this at runtime.

Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: @lhutton1 Thank you very much for clarifications. I am resolving this.

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 2 above)
Comment (mbaret) [reply python/tvm/relay/op/contrib/arm_compute_lib.py:216]: I can still see the change here

**Review by comaniac — CHANGES_REQUESTED** [review]
I'll block it first due to the unresolved comments. @d-smirnov please address the concern raised by @mbaret.

**Review by comaniac — APPROVED** [review]

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 2 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py]: Done. Duplication reverted.

**Comment by comaniac** [comment]
@mbaret PTAL and approve and request changes explicitly. Thanks.

**Review by mbaret — APPROVED** [review]
lgtm now

**Review by comaniac — APPROVED** [review]

**Comment by comaniac** [comment]
Thanks @d-smirnov @mbaret @lhutton1 @trevor-m
